src/algebra/associated.lean

```diff
@@ -129,6 +129,27 @@ lemma prime.pow_dvd_of_dvd_mul_right
   {p a b : α} (hp : prime p) (n : ℕ) (h : ¬p ∣ b) (h' : p ^ n ∣ a * b) : p ^ n ∣ a :=
 by { rw [mul_comm] at h', exact hp.pow_dvd_of_dvd_mul_left n h h' }
 
+lemma prime.dvd_of_pow_dvd_pow_mul_pow_of_square_not_dvd [cancel_comm_monoid_with_zero α]
+  {p a b : α} {n : ℕ} (hn : 0 < n) (hp : prime p) (hpow : p ^ n ∣ a ^ n * b ^ (n - 1))
```

Comment:
How about:
```suggestion
  {p a b : α} (n : ℕ) (hp : prime p) (hpow : p ^ n.succ ∣ a ^ n.succ * b ^ n)
```
which drops a hypothesis and avoids the evil `nat.sub`?

Reply:
Or maybe even:
```lean
{p a b : α} {m n : ℕ} (hmn : m < n) (hp : prime p) (hpow : p ^ n ∣ a ^ n * b ^ m)
```
(optionally with a stronger conclusion)

Reply:
I agree avoiding `nat.sub` is a good idea, but I am not sure I understand which statement you have in mind when you say "a stronger conclusion".

Reply:
Ah, you mean divisibility by higher power of `p`! Let me see.

Reply:
I don't think we can get divisibility by higher powers of `p` (take `b=1)`, and having `b^m` instead of `b^n` is a weaker assumption, so I would stick with the current formulation.